File: packages/py/src/web3_agent_reputation/chains.py
"""7-chain ERC-8004 registry table.

Provenance: copied from ``packages/ts/src/chains/config.ts`` at the `wp-c` branch point
off `develop` — see ``abi/SOURCE.md`` for the full derivation chain (CREATE2 vanity-salt
deployment, ``eth_getCode`` binary search per chain). Not re-derived or re-verified here.
"""

from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class ChainConfig:
    chain_id: int
    name: str
    identity: str
    reputation: str
    validation: str
    # First block at which the registries' bytecode is present on this chain.
    deployment_block: int


# Same CREATE2 vanity-salt deployment across every mainnet chain.
_IDENTITY = "0x8004A169FB4a3325136EB29fA0ceB6D2e539a432"
_REPUTATION = "0x8004BAa17C55a88189AE136b182e5fdA19dE9b63"
_VALIDATION = "0x8004Cc8439f36fd5F9F049D9fF86523Df6dAAB58"

_CHAIN_CONFIGS: dict[int, ChainConfig] = {
    1: ChainConfig(1, "ethereum", _IDENTITY, _REPUTATION, _VALIDATION, 24339871),
    8453: ChainConfig(8453, "base", _IDENTITY, _REPUTATION, _VALIDATION, 41663783),
    137: ChainConfig(137, "polygon", _IDENTITY, _REPUTATION, _VALIDATION, 82458484),
    42161: ChainConfig(42161, "arbitrum", _IDENTITY, _REPUTATION, _VALIDATION, 428895443),
    10: ChainConfig(10, "optimism", _IDENTITY, _REPUTATION, _VALIDATION, 147514947),
    56: ChainConfig(56, "bnb", _IDENTITY, _REPUTATION, _VALIDATION, 79027268),
    100: ChainConfig(100, "gnosis", _IDENTITY, _REPUTATION, _VALIDATION, 44505010),
}


def get_chain_config(chain_id: int) -> ChainConfig | None:
    return _CHAIN_CONFIGS.get(chain_id)


def supported_chain_ids() -> list[int]:
    return list(_CHAIN_CONFIGS.keys())

File: packages/rs/src/chains.rs
//! 7-chain ERC-8004 registry table.
//!
//! Copied byte-for-byte from `packages/ts/src/chains/config.ts` in this same repository
//! — see `../registry/abi/SOURCE.md` for the full derivation provenance
//! (CREATE2 vanity-salt deployment, `eth_getCode` binary search per chain). Not
//! re-derived or re-verified here.

use alloy::primitives::{address, Address};

#[derive(Debug, Clone, Copy)]
pub struct Registries {
    pub identity: Address,
    pub reputation: Address,
    pub validation: Address,
}

#[derive(Debug, Clone, Copy)]
pub struct ChainConfig {
    pub chain_id: u64,
    pub name: &'static str,
    pub registries: Registries,
    /// First block at which the registries' bytecode is present on this chain.
    pub deployment_block: u64,
}

// Same CREATE2 vanity-salt deployment across every mainnet chain.
const IDENTITY: Address = address!("8004A169FB4a3325136EB29fA0ceB6D2e539a432");
const REPUTATION: Address = address!("8004BAa17C55a88189AE136b182e5fdA19dE9b63");
const VALIDATION: Address = address!("8004Cc8439f36fd5F9F049D9fF86523Df6dAAB58");

const REGISTRIES: Registries = Registries {
    identity: IDENTITY,
    reputation: REPUTATION,
    validation: VALIDATION,
};

const CHAIN_CONFIGS: &[ChainConfig] = &[
    ChainConfig {
        chain_id: 1,
        name: "ethereum",
        registries: REGISTRIES,
        deployment_block: 24339871,
    },
    ChainConfig {
        chain_id: 8453,
        name: "base",
        registries: REGISTRIES,
        deployment_block: 41663783,
    },
    ChainConfig {
        chain_id: 137,
        name: "polygon",
        registries: REGISTRIES,
        deployment_block: 82458484,
    },
    ChainConfig {
        chain_id: 42161,
        name: "arbitrum",
        registries: REGISTRIES,
        deployment_block: 428895443,
    },
    ChainConfig {
        chain_id: 10,
        name: "optimism",
        registries: REGISTRIES,
        deployment_block: 147514947,
    },
    ChainConfig {
        chain_id: 56,
        name: "bnb",
        registries: REGISTRIES,
        deployment_block: 79027268,
    },
    ChainConfig {
        chain_id: 100,
        name: "gnosis",
        registries: REGISTRIES,
        deployment_block: 44505010,
    },
];

pub fn get_chain_config(chain_id: u64) -> Option<ChainConfig> {
    CHAIN_CONFIGS
        .iter()
        .copied()
        .find(|c| c.chain_id == chain_id)
}

pub fn supported_chain_ids() -> Vec<u64> {
    CHAIN_CONFIGS.iter().map(|c| c.chain_id).collect()
}
